src/Player.ts

```
@@ -1752,4 +1769,75 @@ export class Player {
+          return new BeginnerCorporation();
+      }
+      return ALL_CORPORATION_CARDS.find((card) => card.name === cardName) 
+      || ALL_PRELUDE_CORPORATIONS.find((card) => card.name === cardName) 
+      || ALL_VENUS_CORPORATIONS.find((card) => card.name === cardName) 
+      || ALL_COLONIES_CORPORATIONS.find((card) => card.name === cardName) 
+      || ALL_PROMO_CORPORATIONS.find((card) => card.name === cardName) 
+      || ALL_TURMOIL_CORPORATIONS.find((card) => card.name === cardName);
+    }
+
+    // Function to return a card object by its name
+    private getProjectCardByName(cardName: string): IProjectCard | undefined {
+      return ALL_PROJECT_CARDS.find((card) => card.name === cardName) 
+      || ALL_PRELUDE_CARDS.find((card) => card.name === cardName) 
+      || ALL_PRELUDE_PROJECTS_CARDS.find((card) => card.name === cardName)
+      || ALL_VENUS_PROJECTS_CARDS.find((card) => card.name === cardName)
+      || ALL_COLONIES_PROJECTS_CARDS.find((card) => card.name === cardName);
+    }
+
+    // Function used to rebuild each objects
+    public loadFromJSON(d: Player): Player {
+      // Assign each attributes
+      let o = Object.assign(this, d);
+
+      // Rebuild generation played map
+      this.generationPlayed = new Map<string, number>();
+      d.generationPlayed.forEach((element: any) => {
+        this.generationPlayed.set(element[0], element[1]);
+      });
+
+      // action this generation set
+      this.actionsThisGeneration = new Set<string>();
+      d.actionsThisGeneration.forEach((element: any) => {
+        this.actionsThisGeneration.add(element);
+      });
+
+      // Rebuild corporation card
+      this.corporationCard = this.getCorporationCardByName(d.corporationCard!.name)
+
+      // Rebuild deal corporation array
+      this.dealtCorporationCards = d.dealtCorporationCards.map((element: CorporationCard)  => {
+        return this.getCorporationCardByName(element.name)!;
```

If for whatever reason these functions return `undefined` we should throw an error of some kind instead of pushing on `undefined`. When I was doing my re-factor yesterday for changing cards to use factories it was very helpful in the other places where we throw instead of using `!`. We should avoid using `!` and `any` as much as possible.